I2C get_config API

niedzwiecki-dawid opened the pull request.

niedzwiecki-dawid committed:
i2c: Add get_config to I2C API

Some applications need to get the current I2C configuration. Add a
proper callback to I2C API under Kconfig option not to change
applications that don't need this feature.

Signed-off-by: Dawid Niedzwiecki <dn@semihalf.com>

niedzwiecki-dawid committed:
i2c: npcx: Add get_config function

Add get_config function to NPCX I2C driver. The master mode is hardcoded
and get the speed from a controller.

Signed-off-by: Dawid Niedzwiecki <dn@semihalf.com>

keith-zephyr commented on include/drivers/i2c.h:
```diff
@@ -553,6 +560,29 @@ static inline int z_impl_i2c_slave_driver_unregister(const struct device *dev)
 	return api->driver_unregister(dev);
 }
 
+#ifdef CONFIG_I2C_GET_CONFIG
```
Can you move this adjacent to i2c_configure()?

semihalf-niedzwiecki-dawid replied:
Done

keith-zephyr commented on include/drivers/i2c.h:
```diff
@@ -553,6 +560,29 @@ static inline int z_impl_i2c_slave_driver_unregister(const struct device *dev)
 	return api->driver_unregister(dev);
 }
 
+#ifdef CONFIG_I2C_GET_CONFIG
```
Can you move this adjacent to the i2c_configure() prototype?

semihalf-niedzwiecki-dawid replied:
Done

keith-zephyr commented on include/drivers/i2c.h:
```diff
@@ -553,6 +560,29 @@ static inline int z_impl_i2c_slave_driver_unregister(const struct device *dev)
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General input / output error.
+ * @retval -ERANGE Out of supported i2c frequency.
```
This would be clearer as "-ERANGE Configured I2C frequency is invalid."

semihalf-niedzwiecki-dawid replied:
Done

de-nordic commented:
Shouldn't be there handler provided, in `i2c_handlers.c`,  for use with `CONFIG_USERSPACE`?

de-nordic commented on drivers/i2c/i2c_npcx_port.c:
```diff
@@ -108,6 +108,28 @@ static int i2c_npcx_port_transfer(const struct device *dev,
+	int ret;
+
+	if (config->i2c_ctrl == NULL) {
+		LOG_ERR("Cannot find i2c controller on port%02x!",
+								config->port);
```
130 characters are allowed, these lines look strange.

semihalf-niedzwiecki-dawid replied:
Done.

sjg20 commented on drivers/i2c/Kconfig:
```diff
@@ -128,4 +128,9 @@ config I2C_RV32M1_LPI2C
+config I2C_GET_CONFIG
+	bool "Enable get configuration API"
+	help
+	  This enables the API to get the current configuration of I2C.
```
Please add more details, at least three lines. Here you could talk about what types of configuration can be obtained

nashif replied:
why do we need this kconfig? it is not common to guard drivers APIs with such kconfig, especially for something that is generic like this

de-nordic replied:
> 
> 
> why do we need this kconfig? it is not common to guard drivers APIs with such kconfig, especially for something that is generic like this

The question is whether we want to increase API size with calls that are not used often. Taking into account that I2C is stable and the function was not needed, so far, this may not be used too often.
Maybe we should consider opposite approach: provide Kconfig to remove API, in case when sb wants to cut out what is not used.

semihalf-niedzwiecki-dawid replied:
I think the current approach is better because it doesn't force users to add additional configs to stay with the stable API. The same approach is in UART API, e.g. CONFIG_UART_LINE_CTRL or CONFIG_UART_DRV_CMD.

sjg20 commented on include/drivers/i2c.h:
```diff
@@ -163,6 +163,8 @@ struct i2c_slave_config;
 typedef int (*i2c_api_configure_t)(const struct device *dev,
 				   uint32_t dev_config);
+typedef int (*i2c_api_get_config_t)(const struct device *dev,
+				    uint32_t *dev_config);
```
Please add a full function comment, including describing what dev_config is

semihalf-niedzwiecki-dawid replied:
It is done under i2c_config_get, as the other API functions.

de-nordic commented on include/drivers/i2c.h:
```diff
@@ -355,6 +362,29 @@ static inline int z_impl_i2c_configure(const struct device *dev,
+
+static inline int z_impl_i2c_get_config(const struct device *dev, uint32_t *dev_config)
+{
+	const struct i2c_driver_api *api =
```
Wouldn't this fit in 130 characters?

semihalf-niedzwiecki-dawid replied:
Yes, done.

de-nordic commented:
I am not sure about the api call name: shouldn't it be rather `get_configuration`, `configuration_get`, or just `configuration`?
By looking at the `get_config` , as e getter,I would expect that the configuration setting function would be called `set_config` but it is named `configure`.

sjg20 commented:
> I am not sure about the api call name: shouldn't it be rather `get_configuration`, `configuration_get`, or just `configuration`? By looking at the `get_config` , as e getter,I would expect that the configuration setting function would be called `set_config` but it is named `configure`.

I think it would be better to rename 'configure' to set_config", then

de-nordic commented:
> 
> I think it would be better to rename 'configure' to set_config", then

@sjg20 Agree: provide `set_config` as `configure` alternative, and set to deprecate the later. Inclusion of `get_config` is already stable API change, we can go with the other change too.

semihalf-niedzwiecki-dawid commented:
> Shouldn't be there handler provided, in `i2c_handlers.c`, for use with `CONFIG_USERSPACE`?

Yes, done.

semihalf-niedzwiecki-dawid commented:
> > I think it would be better to rename 'configure' to set_config", then
> 
> @sjg20 Agree: provide `set_config` as `configure` alternative, and set to deprecate the later. Inclusion of `get_config` is already stable API change, we can go with the other change too.

First of all, I don't want to change the `configure` function because it is naming used by other APIs e.g. uart_configure, i2c_configure, can_configure, etc. 

Regarding how to name the new function, I used `config_get` because a similar function in the UART API is `uart_config_get`.

WealianLiao commented on drivers/i2c/i2c_npcx_port.c:
```diff
@@ -108,6 +108,27 @@ static int i2c_npcx_port_transfer(const struct device *dev,
+
+	ret = npcx_i2c_ctrl_get_speed(config->i2c_ctrl, &speed);
+	if (!ret) {
+		*dev_config = (I2C_MODE_MASTER | I2C_ADDR_10_BITS | speed);
```
Don't support I2C_ADDR_10_BITS

semihalf-niedzwiecki-dawid replied:
Done.

WealianLiao replied:
The commit log for 10-bits might be confused. Let's fix it.

semihalf-niedzwiecki-dawid replied:
Yes, thanks!

de-nordic commented:
@semihalf-niedzwiecki-dawid I think the change requires following https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html

niedzwiecki-dawid committed:
i2c: emul: Add get_config function

Add get_config function to I2C emulator.

Also update tests using I2C emulator to use i2c_get_config.

Signed-off-by: Dawid Niedzwiecki <dn@semihalf.com>

galak reviewed (dismissed):
please add a test to tests/drivers/i2c for this new API.

niedzwiecki-dawid committed:
i2c: tests: update i2c_api test

Add testing of the new i2c_get_config function to the i2c_api test.

Signed-off-by: Dawid Niedzwiecki <dn@semihalf.com>

carlescufi commented:
@galak and @sjg20 could you please take another look?
